java/src/main/native/ai_onnxruntime_OrtTrainingSession.c

```diff
@@ -0,0 +1,745 @@
+  (void)jobj;  // Required JNI parameter not needed by functions which don't need to access their host object.
+  const OrtApi* api = (const OrtApi*)apiHandle;
+  const OrtTrainingApi* trainApi = (const OrtTrainingApi*)trainApiHandle;
+  OrtTrainingSession* trainSession = (OrtTrainingSession*)nativeHandle;
+  const OrtRunOptions* options = (const OrtRunOptions*) runOptionsHandle;
+  checkOrtStatus(jniEnv, api, trainApi->OptimizerStep(trainSession, options));
+}
+
+/*
+ * Class:     ai_onnxruntime_OrtTrainingSession
+ * Method:    registerLinearLRScheduler
+ * Signature: (JJJJJF)V
+ */
+JNIEXPORT void JNICALL Java_ai_onnxruntime_OrtTrainingSession_registerLinearLRScheduler
+  (JNIEnv * jniEnv, jobject jobj, jlong apiHandle, jlong trainApiHandle, jlong nativeHandle, jlong warmupSteps, jlong totalSteps, jfloat initialLearningRate) {
+  (void)jobj;  // Required JNI parameter not needed by functions which don't need to access their host object.
+  const OrtApi* api = (const OrtApi*)apiHandle;
+  const OrtTrainingApi* trainApi = (const OrtTrainingApi*)trainApiHandle;
+  OrtTrainingSession* trainSession = (OrtTrainingSession*)nativeHandle;
+  checkOrtStatus(jniEnv, api, trainApi->RegisterLinearLRScheduler(trainSession, warmupSteps, totalSteps, initialLearningRate));
+}
+
+/*
+ * Class:     ai_onnxruntime_OrtTrainingSession
+ * Method:    schedulerStep
+ * Signature: (JJJ)V
+ */
+JNIEXPORT void JNICALL Java_ai_onnxruntime_OrtTrainingSession_schedulerStep
+    (JNIEnv * jniEnv, jobject jobj, jlong apiHandle, jlong trainApiHandle, jlong nativeHandle) {
+  (void)jobj;  // Required JNI parameter not needed by functions which don't need to access their host object.
+  const OrtApi* api = (const OrtApi*)apiHandle;
+  const OrtTrainingApi* trainApi = (const OrtTrainingApi*)trainApiHandle;
+  OrtTrainingSession* trainSession = (OrtTrainingSession*)nativeHandle;
+  checkOrtStatus(jniEnv, api, trainApi->SchedulerStep(trainSession));
+}
+
+/*
+ * Class:     ai_onnxruntime_OrtTrainingSession
+ * Method:    exportModelForInference
+ * Signature: (JJJJLjava/lang/String;[Ljava/lang/String;)V
+ */
+JNIEXPORT void JNICALL Java_ai_onnxruntime_OrtTrainingSession_exportModelForInference
+  (JNIEnv * jniEnv, jobject jobj, jlong apiHandle, jlong trainApiHandle, jlong nativeHandle, jstring outputPath, jlong numOutputs, jobjectArray outputNamesArr) {
+  (void)jobj;  // Required JNI parameter not needed by functions which don't need to access their host object.
+  const OrtApi* api = (const OrtApi*)apiHandle;
+  const OrtTrainingApi* trainApi = (const OrtTrainingApi*)trainApiHandle;
+  OrtTrainingSession* trainSession = (OrtTrainingSession*)nativeHandle;
+
+  // prep output names array
+  const char** outputNames = malloc(sizeof(char*) * numOutputs);
+  if (outputNames == NULL) {
+    return;
+  }
+  jobject* javaOutputStrings = malloc(sizeof(jobject) * numOutputs);
+  if (javaOutputStrings == NULL) {
+    free(outputNames);
```

Comment:
> free(outputNames); [](http://example.com/codeflow?start=3&length=19)

Should we continue here?

Reply:
Good point, it should throw an error message.